### PR TITLE
feat: command line tool to copy source files from googleapis-gen to a…

### PR DESCRIPTION
**Pull request by SurferJeffAtGoogle**
… library repo

### CODE REVIEW

File: packages/owl-bot/src/copy-code.ts [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+import {
+  owlBotYamlPath,
+  owlBotYamlFromText,
+  OwlBotYaml,
+  toFullMatchRegExp,
+} from './config-files';
+import path from 'path';
+import {v4 as uuidv4} from 'uuid';
+import * as fs from 'fs';
+import * as fse from 'fs-extra';
```
[the remark this answers is not in the document]
Comment (bcoe) [reply packages/owl-bot/src/copy-code.ts:27]: FYI, we're working on pulling recursive copy from fs-extra into Node core.

File: packages/owl-bot/test/copy-code.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import {OwlBotYaml} from '../src/config-files';
+
+describe('copyDirs', () => {
+  /**
+   * Creates a sample source tree.
+   */
+  function makeSourceTree(rootDir: string): string {
+    const dirs = [
+      'source',
+      'source/a',
```
Comment (bcoe) [inline packages/owl-bot/test/copy-code.ts:31]: add an `after` helper that cleans up the directories?

**Review by bcoe — CHANGES_REQUESTED** [review]
Caught a few typos, and left some suggestions for `TODO` comments. Most notably:

* I'm fine with us using `sync` helpers, but we should add warnings to any helpers that do, so we don't use them in our web facing code (as it will lock requests).
* our CLI is functional now, so we'll probably want to start fleshing out some happy path tests for them (not a blocker for this PR).

File: packages/owl-bot/src/copy-code.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/copy-code.ts:27]: Yay.  Very useful.

File: packages/owl-bot/src/copy-code.ts [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+      const fullSourcePath = path.join(sourceDir, sourcePath);
+      const relPath = ('/' + sourcePath).replace(regExp, deepCopy.dest);
+      const fullDestPath = path.join(destDir, relPath);
+      const dirName = path.dirname(fullDestPath);
+      if (!stat(dirName)?.isDirectory()) {
+        logger.info('mkdir ' + dirName);
+        fs.mkdirSync(dirName, {recursive: true});
+      }
+      logger.info(`cp -r ${fullSourcePath} ${fullDestPath}`);
+      fse.copySync(fullSourcePath, fullDestPath, {
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/copy-code.ts:297]: Adding a comment to the top of copy-code.ts.

File: packages/owl-bot/src/copy-code.ts [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+ * @param destRepo the repo to search
+ * @param sourceCommitHash the string to search for
+ */
+export async function copyExists(
+  octokit: OctokitType,
+  destRepo: string,
+  sourceCommitHash: string,
+  logger = console
+): Promise<boolean> {
+  const q = `repo:${destRepo}+${sourceCommitHash}`;
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/copy-code.ts:318]: As I described in our meeting, search is for older issues and PRs.  Code on lines 331 through 354 looks at the most recent 100 issues and pull requests.

File: packages/owl-bot/src/copy-code.ts [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+      fse.copySync(fullSourcePath, fullDestPath, {
+        recursive: true,
+        overwrite: true,
+      });
+    }
+  }
+}
+
+/**
+ * Searches for instances of the sourceCommitHash in recent pull requests and commits.
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/copy-code.ts]: Actually, it's accurate.  I added a comment describing the return value.

File: packages/owl-bot/src/octokit-util.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  | InstanceType<typeof Octokit>
+  | InstanceType<typeof ProbotOctokit>;
+
+export interface OctokitParams {
+  'pem-path': string;
+  'app-id': number;
+  installation: number;
+}
+
+export async function octokitFrom(argv: OctokitParams): Promise<OctokitType> {
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/octokit-util.ts:38]: I added a TODO.

File: packages/owl-bot/test/copy-code.ts (hunk 2 above)
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/test/copy-code.ts:31]: They're created in the temp directory, and the OS clears out the temp directory on a regular schedule.

File: packages/owl-bot/src/bin/commands/copy-code.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        describe:
+          'The directory containing a local repo.  Example: nodejs/vision.',
+        type: 'string',
+        demand: false,
+      });
+  },
+  async handler(argv) {
+    const destDir = argv.dest ?? process.cwd();
+    await copyCode(
+      argv['source-repo'],
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/bin/commands/copy-code.ts:54]: Yep.  Tests are due in a follow up PR.

**Comment by SurferJeffAtGoogle** [comment]
Ready for another look.

**Review by bcoe — APPROVED** [review]

**Pull request merged**